src/v0.4/parseExpr_staged.jl

```diff
@@ -305,32 +305,16 @@ end
 @generated addToExpression_reorder(ex, arg) = :(addToExpression(ex, 1.0, arg))
 
 @generated function addToExpression_reorder(ex, args...)
-    if !isleaftype(ex) || mapreduce(t -> !isleaftype(t), |, args)
-        error("Can't process abstract types")
-    end
-    # how many of the multiplicands are variables?
-    n_var = mapreduce(t -> (t == Variable || t == AffExpr), +, args)
-    has_quad = mapreduce(t -> (t == QuadExpr), |, args)
-    if n_var == 0 && !has_quad
-        #println("No variables")
-        return :(addToExpression(ex, 1.0, (*)(args...)))
-    elseif n_var == 1 && !has_quad # linear
-        #println("Linear")
-        coef_expr = Expr(:call, :*)
-        coef_idx = Int[]
-        var_idx = 0
-        for i in 1:length(args)
-            if args[i] == Variable || args[i] == AffExpr
-                var_idx = i
-            else
-                push!(coef_expr.args, :(args[$i]))
-            end
-        end
-        return :(addToExpression(ex, $coef_expr, args[$var_idx]))
+    n = length(args)
+    varidx = find(t -> (t == Variable || t == AffExpr), collect(args))
+    allscalar = all(t -> (t <: Number), args[setdiff(1:n, varidx)])
+    if allscalar && length(varidx) == 1
+        idx = varidx[1]
+        @show :(addToExpression(ex, (*)(args[setdiff(1:$n,$idx)]...), args[$idx]))
+    elseif n == 2 # special-case to avoid calls to unary multiplication
+        :(addToExpression(ex, args[1], args[2]))
     else
-        #println("Nonlinear")
-        coef_expr = Expr(:call, :*, [:(args[$i]) for i in 1:(length(args)-1)]...)
-        return :(addToExpression(ex, $coef_expr, args[$(length(args))]))
+        :(addToExpression(ex, (*)(args[1:end-1]...), args[end]))
```

Comment:
I think this form ends up splatting at runtime instead. That could be a big performance issue